src/python_testing/TC_DGSW_2_3.py

```diff
@@ -0,0 +1,201 @@
+        self.step(5)
+        if self.pics_guard(Clusters.SoftwareDiagnostics.Attributes.CurrentHeapUsed.attribute_id in attribute_list):
+            current_heap_used_original = await self.read_dgsw_attribute_expect_success(endpoint=endpoint, attribute=attributes.CurrentHeapUsed)
+            asserts.assert_true(self.is_valid_uint64_value(current_heap_used_original),
+                                "CurrentHeapUsed field should be a uint64 type")
+            if high_watermark_original is not None:
+                asserts.assert_true(current_heap_used_original <= high_watermark_original,
+                                    "CurrentHeapUsed should be less than or equal to CurrentHeapHighWatermark")
+
+        # STEP 6: TH sends to the DUT the  command
+        self.step(6)
+        response = await self.send_reset_watermarks_command()
+        logging.info(f"ResetWatermarks response: {response}")
+
+        # STEP 7: TH reads from the DUT the CurrentHeapHighWatermark attribute
+        self.step(7)
+        if self.pics_guard(Clusters.SoftwareDiagnostics.Attributes.CurrentHeapHighWatermark.attribute_id in attribute_list):
+            current_heap_high_watermark = await self.read_dgsw_attribute_expect_success(endpoint=endpoint, attribute=attributes.CurrentHeapHighWatermark)
+            asserts.assert_true(self.is_valid_uint64_value(current_heap_high_watermark),
+                                "CurrentHeapHighWatermark field should be a uint64 type")
+
+            # Verify that the returned value is <= high_watermark_original
+            asserts.assert_true(current_heap_high_watermark <= high_watermark_original,
+                                f"CurrentHeapHighWatermark ({current_heap_high_watermark}) should be <= high_watermark_original ({high_watermark_original})")
+
+            # If CurrentHeapUsed is supported
+            if current_heap_used_original is not None and current_heap_used_original < high_watermark_original:
+                # Verify that the returned value is < high_watermark_original
+                asserts.assert_true(current_heap_high_watermark < high_watermark_original,
+                                    f"CurrentHeapHighWatermark ({current_heap_high_watermark}) should be < high_watermark_original ({high_watermark_original}) when CurrentHeapUsed ({current_heap_used_original}) is less than high_watermark_original")
+
+        # STEP 8: TH reads from the DUT the ThreadMetrics attribute
+        self.step(8)
+        if self.pics_guard(Clusters.SoftwareDiagnostics.Attributes.ThreadMetrics.attribute_id in attribute_list):
+            thread_metrics_reset = await self.read_dgsw_attribute_expect_success(endpoint=endpoint, attribute=attributes.ThreadMetrics)
+            # the Id field is mandatory
+            asserts.assert_true(self.is_valid_uint64_value(thread_metrics_reset[0].id), "Id field should be a uint64 type")
+            if thread_metrics_reset[0].name is not None:
```

Comment:
test plan says `For each item in the list, if there is a...` so it seems like we should test the entire list rather than only element 0